streamalert/rules_engine/rules_engine.py

```diff
@@ -213,15 +213,82 @@ def _rule_analysis(self, payload, rule):
+            rule (rule.Rule): Attributes for the rule which triggered the alert
+        Returns:
+            set: unique set of outputs, only required outputs if the rule is staged
+        """
         # Check if the rule is staged and, if so, only use the required alert outputs
         if rule.is_staged(self._rule_table):
             all_outputs = self._required_outputs_set
         else:  # Otherwise, combine the required alert outputs with the ones for this rule
             all_outputs = self._required_outputs_set.union(rule.outputs_set)
+            if rule.dynamic_outputs:
+                dynamic_outputs = self._configure_dynamic_outputs(record, rule)
+                all_outputs = all_outputs.union(dynamic_outputs)
 
         return all_outputs
 
+    @classmethod
+    def _configure_dynamic_outputs(cls, record, rule):
+        """Generate list of outputs from dynamic_outputs
+
+        Args:
+            record (dict): Record to pass through to the dynamic_output function
+            rule (rule.Rule): Attributes for the rule which triggered the alert
+        Returns:
+            list: list of additional outputs to append to the current set
+        """
+        args_list = [record]
+        if rule.context:
+            # Pass context to dynamic_output function if context exists
+            args_list.append(rule.context)
+
+        additional_outputs = []
+        for dynamic_output in rule.dynamic_outputs_set:
+            outputs = cls._call_dynamic_output(dynamic_output, rule.name, args_list)
+            if not outputs:
+                continue
+
+            for output in outputs:
+                if ":" not in output:
+                    # Log a warning and skip output if : is not in the output
+                    LOGGER.warning(
+                        "Invalid output configured by %s: %s missing ':'",
+                        dynamic_output.__name__, output
+                    )
+                    continue
+
+                additional_outputs.append(output)
+
+        return additional_outputs
+
+    @staticmethod
+    def _call_dynamic_output(dynamic_output, rule_name, args_list):
+        """Call the dynamic_output function
+
+        Args:
+            dynamic_output (func): Callable function which returns None, str or List[str]
+            rule_name (str): The name of the rule the functions belong to
+            args_list (list): list of args to be passed to the dynamic function
+                should be (record or record and context)
+        Returns:
+            list: list of additional outputs
+        """
+        try:
+            outputs = dynamic_output(*args_list)
+        except Exception:  # pylint: disable=broad-except
+            LOGGER.error(
+                "Exception when calling dynamic_output %s for rule %s",
+                dynamic_output.__name__, rule_name
+            )
+        else:
+            return outputs if isinstance(outputs, list) else [outputs]
```

Comment:
In the case `dynamic_output()` returns `None`, this will actually return a list of a single `None`, which I don't think we want

Reply:
i'll potentially set something like:

## My preferred option
```python
else:
    if outputs:
       # return outputs otherwise just default to None
        return outputs if isinstance(outputs, list) else [outputs]
```

or 

## second option (less clean, alot of return statements)
```python
else:
    if not outputs:
        # Return if outputs are None
        return
    return outputs if isinstance(outputs, list) else [outputs]
```

Reply:
I'm a fan of the 2nd (I like early returns).

Also, it should return empty array, ya? To be consistent with the advertised return type in the docblock

Reply:
also a fan of this version - good change :)